### PR TITLE
Update Procedures Functionality

### PR DESCRIPTION
**Pull request by sjpadgett**
1. New feature to allow adding custom groups of favorite orders.
  a.  Add group feature in configuration
  b. imports all orders in favorite group from configuration into form.
  c. added favorite search to order picker.
  d. can give a default icd code to favorite order.
2. Procedure form bug fixes.
3. New delete procedure order items button in forms order list.
4. Updated QOE dependencies, calendar and formatting.
5. List reports for electronic results now will help resolve orphaned orders. (missing order or patient).

Below added by bradymiller:
Can test on following Up For Grabs demo:
https://www.open-emr.org/wiki/index.php/Development_Demo#Beta_-_Up_For_Grabs_Demo

### CODE REVIEW

**Comment by sjpadgett** [comment]
This PR is so awesome that travis simply fainted :) @bradymiller I hope I didn't miss anything but a docker w/data may be wise and hopefully a couple folks will help test with me. Was that subtle @zbig01  and @stephenwaite :)

**Comment by zbig01** [comment]
@sjpadgett looking at your latest PR, can you tell me how the Custom Favorite Group and Custom Favorite Item features are activated, more specifically where in the hierarchy does it figure. I note that it has an Identifying Code so I guess it must be above Procedure Order.
Thanks ;)

**Comment by sjpadgett** [comment]
They are handle in configuration just as if they are a normal group or order. So you need to do nothing with configuration. The only difference is how they are flagged so form and picker know what they are.
To activate you have to add the list options to procedure types in Administration->Lists. Take a look at the database/501 to 502 sql table updates in this PR to get values.
The favorites are handled in the procedure form and order picker. About the only thing you may want to do is adjust the tips and help.

**Comment by zbig01** [comment]
@sjpadgett 
I created a Custom Favorite Group under a group called it preop labs , included 2 custom favorite items CBC and Renal Panel, gave it a ICD10 of C20 and saved it. I am able to place an order and they both show up, note the ICD10 does not populate and the Renal Panel does not have the AOE section.

jery-procedure-order 1

**Comment by sjpadgett** [comment]
ICD works for me. On new order items the Question header doesn't show until saved, however, if the new order have questions when selected from picker then they will populate for entry. It is done this way by design because it gives an indication the orders have not been saved.

**Comment by zbig01** [comment]
@sjpadgett 
quick question what happens to the Identifying Code if a test is included as a Custom Favorite Item as well as a General test, say for example I need CBC as a CFI as well as a Procedure Order what Identifying Code will be used for each instance of the test  and how will it figure in a HL7 message

**Comment by sjpadgett** [comment]
You can add/remove ICD's in the form. Most times a default ICD will not be used in configuration but some folks do. The ICD10 should populate in the form when you select the favorite from the picker if it is populated in the favorite item under favorite group in configuration. If not working for you then I may have missed something in PR because it works for me and has been for days. Also, for now, the only place a favorite order works is when is a part of a favorite group. So adding a favorite order to a normal group is not supported.
You have to remember that the HL7 for the order comes from the procedure form so any codes that are sent are sent from what is in the form, not the compendium. The compendium merely supplies it to the form. The identifying code is normally the code the lab uses to decide what test is being ordered. However, I think it would be useful to display the ordering code in the form. As of now it is hidden.

**Comment by zbig01** [comment]
I think I did not state my question clearly so let me try again. Say I create a CBC order as a CFI and I create another separate CBC order as a Procedure Order that is not a part of the CFG but in a whole different Group will they both have the same Identifying Code as they both are hypothetically being sent to the same lab. 
If so does the dictum that Identifying Codes be unique still hold true when it comes to displaying results as in the past I had noted that if the identifying Code was left blank you would get a sort of Cartesian Join with all other Procedure Orders that had blank Identifying codes

**Comment by sjpadgett** [comment]
No I understood the question. I just don't know how to answer any differently.  Also, I missed an edit in this PR because the ICD's are not saving after a fresh install from this PR version. So, you're right there. For the configuration questions I think a phone call tomorrow or Mon. will make hashing that out easier if you're willing.

**Comment by zbig01** [comment]
Sure. We can talk tomorrow. I will work with the AOE part and check it out. Tx

File: interface/forms/procedure_order/new.php [hunk 1]
```diff
@@ -332,77 +377,94 @@ function addProcLine() {
...
- rcvarname = varname;
- // codetype is just to make things easier and avoid mistakes.
- // Might be nice to have a lab parameter for acceptable code types.
- // Also note the controlling script here runs from interface/patient_file/encounter/.
- dlgopen('find_code_dynamic.php?codetype=<?php echo attr(collect_codetypes("diagnosis", "csv")); ?>', '_blank', 985, 750, '', '<?php echo xla("Select Diagnosis Codes"); ?>');
+    rcvarname = varname;
+    // codetype is just to make things easier and avoid mistakes.
+    // Might be nice to have a lab parameter for acceptable code types.
+    // Also note the controlling script here runs from interface/patient_file/encounter/.
+    dlgopen('find_code_dynamic.php?codetype=ICD10', '_blank', 825, 750, '', '<?php echo xla("Select Diagnosis Codes"); ?>');
```
Comment (bradymiller) [inline interface/forms/procedure_order/new.php]: You should not need to hardcode ICD10 here. the collect_codetypes will do this (and allow folks to use other diagnosis code types as needed, which are set in administration->Lists->Code Types

File: interface/orders/find_order_popup.php [hunk 2]
```diff
@@ -31,12 +31,19 @@
...
     $typeid = $_GET['typeid'] + 0;
     $name = '';
     if ($typeid) {
-        $ptrow = sqlQuery("SELECT name FROM procedure_type WHERE " .
-            "procedure_type_id = '$typeid'");
+        $ptrow = sqlQuery("SELECT * FROM procedure_type WHERE procedure_type_id = ?", array($typeid));
         $name = addslashes($ptrow['name']);
+        $codes = addslashes($ptrow['related_code']);
+        if ($ptrow['procedure_type'] == 'fgp') {
+            $res = sqlStatement("SELECT * FROM procedure_type WHERE parent = '$typeid' && procedure_type = 'for' ORDER BY seq, name, procedure_type_id");
```
Comment (bradymiller) [inline interface/orders/find_order_popup.php]: use binding

File: interface/orders/find_order_popup.php [hunk 3]
```diff
@@ -109,57 +125,51 @@ function selcode(typeid) {
...
+                    <th><?php echo xlt('Code'); ?></th>
+                    <th><?php echo xlt('Description'); ?></th>
+                    </thead>
+                    <?php
+                    $ord = isset($_REQUEST['bn_search']) ? 'ord' : 'fgp';
+                    $search_term = '%' . $_REQUEST['search_term'] . '%';
+                    $query = "SELECT procedure_type_id, procedure_code, name " .
+                        "FROM procedure_type WHERE " .
+                        "lab_id = ? AND " .
+                        "procedure_type LIKE '$ord' AND " .
```
Comment (bradymiller) [inline interface/orders/find_order_popup.php]: use binding

File: interface/orders/types_edit.php [hunk 4]
```diff
@@ -166,9 +166,14 @@ function del_related(s) {
...
 function sel_related(varname) {
- if (typeof varname == 'undefined') varname = 'form_related_code';
- rcvarname = varname;
- dlgopen('../patient_file/encounter/find_code_dynamic.php', '_blank', 900, 600);
+    if (typeof varname == 'undefined')
+        varname = 'form_related_code';
+    rcvarname = varname;
+    let url = '../patient_file/encounter/find_code_dynamic.php';
+    if (varname == 'form_diagnosis_code')
+        url = '../patient_file/encounter/find_code_dynamic.php?codetype=ICD10';
```
Comment (bradymiller) [inline interface/orders/types_edit.php]: See above comment regarding hard-coding of ICD10

**Comment by bradymiller** [comment]
Just did a code review.

**Comment by bradymiller** [comment]
btw, did a travis rebuild and it didn't crash and burn again :)

**Comment by zbig01** [comment]
Here are some of the scenarios I tried and their results 

Imported the AOE into openEMR - No problems

Created a Custom Favorite Group as a Top Group and included Glucose as a single CFI with Result as Glucose with no problems. I gave it an Identifying Code of 001032

Created a Procedure Order for Alice Brown using this CFI - No Problems, AOE showed up as predicted


jery-procedure-order8


Looked at Procedures > Pending Review, expected a single line for Glucose, found it - No problems

jery-procedure-order3

Created a regular Group - Serum Chemistry with a single Procedure Order Glucose with Identifying Code of 001302 - Immediate problem, The Serum Chemistry group showed a Cartesian Join and showed 2 Glucose Procedure Orders each with an Identifying Code of 001302. Note they appear properly nested with a Tier value of 2

jery-procedure-order2


Created a Discrete Result for the first Glucose in Serum Chemistry with Identifying Code of GR001032, No problem shows a Tier value of 3 as expected. The other Glucose in Serum Chemistry did not show  this Tier as expected.

jery-procedure-order4


Created an order for Glucose Order for Homer Simpson using this Glucose Procedure Order, AOE showed up as expected - No problem

jery-procedure-order6


Looked at Procedures > Pending Review for Homer Simpson - Problem - Noted a Cartesian join, note differing Code values under Results and Recommendations

jery-procedure-order5


Went back to Procedures > Pending Review for Alice Brown - Problem, now showing a Cartesian Join, note differing Code values under Results and Recommendations

jery-procedure-order7

**Comment by bradymiller** [comment]
This PR can be tested on the following demo:
https://www.open-emr.org/wiki/index.php/Development_Demo#Beta_-_Up_For_Grabs_Demo

**Comment by zbig01** [comment]
Screenshot from Demo

jery-procedure-order9


 Note issue with My Custom Group Description

Cartesian join 

jery-procedure-order10

**Comment by sjpadgett** [comment]
Hmm, I never considered custom groups being used for anything but electronic orders where results will be automatically added to report during receive parsing. Therefore, I did not add results to configuration as a third tier. Suppose I should but I just don't see a reason to have results in configuration unless doing orders by hand in which case, one would just add a normal group. I'm thinking custom Profiles or Panels where a user may want to order say 3 or 4 various lab tests that the lab does not offer as a panel but as individual test. Tailored specific drug screens come to mind. Eventually i'll be adding an updated receive parser and more universal compendium loader.

**Comment by zbig01** [comment]
To back this a bit further and think about it as inserting structured data into the database and its subsequent display it should not matter the mechanism of inserting the data,it should be structured no matter how we insert it. The Cartesian display may represent one of two issues, either the data is not structured right or the SQL to extract it is incorrect.

I had grappled with the concept of grouping of tests when I was trying to understand the Procedure Orders module and in fact highlighted it in the help file. To quote my own help file ;)

> Remember grouping of tests is only for conveniently organizing the data in the Procedure Order module in openEMR, for this process to work, configuring the Procedure Orders (the tests themselves) and the Discrete Result (the place to enter and display the returned result) are critically important.


Your new grouping feature is a direct rebuttal to the above statement. Regarding the Custom Favorites group I see your intent i.e. providing a mechanism of artificially grouping tests so that a number of tests can be ordered together, a la panel of tests like Renal Panel etc. This has necessitated creating an Identifying Code for the group and likely using it to display a bunch of orders grouped together. It works fine at the front end i.e. creating the Procedure Order but seems to have some issues at the back end i.e Procedures > Pending Review.

**Comment by zbig01** [comment]
Just received your email will work on the data you supplied :)

**Comment by zbig01** [comment]
1. New feature to allow adding custom groups of favorite orders.
    a. Add group feature in configuration - **Works with No problem**
    b. imports all orders in favorite group from configuration into form. - **Works but default ICD10 not showing up**
    c. added favorite search to order picker. - **Works**
    d. can give a default icd code to favorite order.  - **Not working**
2. Procedure form bug fixes. - **Do not know how to test**
3. New delete procedure order items button in forms order list.  -  **Works**
4. Updated QOE dependencies, calendar and formatting. - **used supplied AOE, do not know how to test for the rest**
5. List reports for electronic results now will help resolve orphaned orders. (missing order or patient). - **still testing** 

In summary most features work as advertised, some kinks need to be ironed out, pending testing of item 5

:+1: :+1:

**Comment by zbig01** [comment]
### Testing HL7 ORU^R01 - Unsolicited transmission of an observation message

Created directory in sites/default/hl7 - **Success**

In Procedures > Providers changed the External lab Protocol to Local Filesystem - **Success**

Changed the Results path to /var/www/html/openemr-master/sites/default/hl7 - **Success**

Created a Procedure Order in Procedures > Configuration ZM04 High Risk Panel 1, no  Discreet Result specified - **Success**

jery-procedure-order17


Placed HL7 ORU file in that directory - **Success**

Opened Procedures > Electronic Reports - **Success**

This is what I get

jery-procedure-order18

Is the error because of write permissions ?

I also note that a dummy New Patient -  Jerry Padgett has been added to openEMR with only DOB - **Surprise** :)

jery-procedure-order11

Procedures > Pending Review - **Success - mostly**

jery-procedure-order13

 Note Missing Performing facility Info

jery-procedure-order14

Procedures > Patient Results - **Success - mostly**

jery-procedure-order15

 Note Missing Performing facility Info
jery-procedure-order19

Procedures > Lab Overview - Success
jery-procedure-order12


jery-procedure-order16


Works almost flawlessly !!  :+1:

**Comment by zbig01** [comment]
I take that back, facility listed under ?

**Comment by zbig01** [comment]
Bootstrapped `/interface/orders/orders_results.php ` whenever you are ready

jery-procedure-order20

Additional info as bootstrapped popup modal - moveable and resizeable

jery-procedure-order21

**Comment by sjpadgett** [comment]
Orphaned orders work if there is not an existing order against the result. You will get a new button in that case to select a patient to apply the order against. You got a patient created because if there is an order to apply the result to then, there should be a patient that order belongs. So. kind of a fluke here.
@zbig01 , really nice job on the types configuration and should be very helpful to new comers to OpenEMR.
@bradymiller I'm done with this 'til I get to adding the generic compendium loader and converting the HL7 stuff to 2.5 and adding a couple more segments. I've tested and will be using in my family's practice so if you haven't any comments then bring this in and Zbig can do his thing and I will keep eye out for bugs.

**Comment by zbig01** [comment]
@sjpadgett thanks for your work. Is it possible you could forward a few simpler HL7 ORU messages. I just want to test a bit more. I just want to try to figure out how the new patient got added without any input.

I note that you are planning to convert v2.3 to 2.5. I saw somewhere that a majority of sites used either 2.3 or 2.3.1. Maybe my info is incorrect.

Instead would it be more adaptable if the MSH header was first checked for the version and the subsequent script deal with the message according to the version?

**Comment by sjpadgett** [comment]
Sure, what I meant but labs are insisting on 2.5 more so these days.

**Comment by bradymiller** [comment]
@sjpadgett , ok, I'll do the honors and bring this in :)

**Comment by zbig01** [comment]
@sjpadgett  @bradymiller  Now that the PR is in I will make style adjustments to 
```
interface/orders/list_reports.php 
interface/orders/orders_results.php
interface/orders/load_compendium.php
```
I will also change the help files accordingly to reflect the part about electronic orders

**Comment by stephenwaite** [comment]
thanks for including me on this @sjpadgett, nice teamwork @zbig01 @bradymiller 
a general search on github for hl7 looking for a php library resulted in dozens of fhir projects with everything but php but did find this starter https://github.com/senaranya/HL7

**Comment by sjpadgett** [comment]
Hi @stephenwaite , I use that library for my HIE stuff but find in most cases it is just as easy to parse in straight code. Also it requires PHP7.1 or greater.
